### PR TITLE
SN1-300 Punish validators who are not setting any weights to sn15

### PR DESCRIPTION
**Pull request by wendellswa06**
## Description

_Fetch inmemory validator hotkeys who are voting sn15 using subtensor apis and based on result whether the valis who query the miner is in the inmemory hotkeys or not, punish them._

## Acceptance Criteria

- [x] _Fetch inmemory validator hotkeys._
- [x] _we should blacklist valdiators which try to use our query synapses, allow all validators to perform discovery/challenge.
- [x] _blacklisting should be without any propability of sucess, so if vali is setting weight then he can use the query endpoint..

## How Has This Been Tested?

- [x] Unit Tests
Fetch inmemory hotkeys successfully.
Punish the non-inmemory hotkeys in query endpoint.

Require test in mainnet as it fetchs inmemory hotkeys from blocks in the mainnet.

### CODE REVIEW

File: neurons/miners/blacklist.py [hunk 1]
```diff
@@ -135,4 +136,6 @@ def base_blacklist(self, synapse: bt.Synapse) -> typing.Tuple[bool, str]:
         return True, f"Blacklisted hotkey: {hotkey}"
     if hotkey not in self.miner_config.whitelisted_hotkeys and self.config.mode == 'prod':
         return True, f"Not Whitelisted hotkey: {hotkey}"
+    if len(self.miner_config.inmemory_hotkeys) >= 1 and hotkey not in self.miner_config.inmemory_hotkeys and random.random() < self.miner_config.vali_penalty_rate:
```
Comment (aph5nt) [inline neurons/miners/blacklist.py:139]: Now please add a condition flag, so we can remotely enable / disable this feature.
+ I need logs who would be punished even if feature is disabled

File: neurons/remote_config.py [hunk 2]
```diff
@@ -85,6 +88,35 @@ def stop_update_thread(self):
         self.stop_event.set()
         self.thread.join()
 
+    def fetch_inmemory_hotkeys(self, current_netuid: int=15):        
+        finney_subtensor = bt.subtensor(network='finney')
```
Comment (aph5nt) [inline neurons/remote_config.py:92]: can we use our subtensor? finney is public one and its not under our control

File: neurons/remote_config.py [hunk 3]
```diff
@@ -50,7 +51,9 @@ def load_remote_config(self):
                     response = requests.get(self.config_url, timeout=10)
                     response.raise_for_status()
                     self.config_cache = response.json()
-
+                                        
+                    self.config_cache.update(self.fetch_inmemory_hotkeys(current_netuid=15))
```
Comment (aph5nt) [inline neurons/remote_config.py:55]: can we use netui from cofig ?

**Review by aph5nt — CHANGES_REQUESTED** [review]
we need few improvements

File: neurons/remote_config.py (hunk 3 above)
Comment (wendellswa06) [reply neurons/remote_config.py:55]: hmm, i think need to add netuid to the load_and_get_config_values. anyway let me try and update.

File: neurons/remote_config.py (hunk 2 above)
Comment (wendellswa06) [reply neurons/remote_config.py:92]: are you planning to use  the subtensor miners are connected

Comment (aph5nt) [reply neurons/remote_config.py:92]: this have to also work on our private network, so we can not use finney in that case

File: neurons/remote_config.py [hunk 4]
```diff
@@ -113,12 +148,15 @@ def load_and_get_config_values(self):
         self.whitelisted_hotkeys = self.get_config_value('whitelisted_hotkeys', ["5FFApaS75bv5pJHfAp2FVLBj9ZaXuFDjEypsaBNc1wCfe52v", "5HK5tp6t2S59DywmHRWPBVJeJ86T61KjurYqeooqj8sREpeN", "5EhvL1FVkQPpMjZX4MAADcW42i3xPSF1KiCpuaxTYVr28sux", "5CXRfP2ekFhe62r7q3vppRajJmGhTi7vwvb2yr79jveZ282w", "5DvTpiniW9s3APmHRYn8FroUWyfnLtrsid5Mtn5EwMXHN2ed", "5F4tQyWrhfGVcNhoqeiNsR6KjD4wMZ2kfhLj4oHYuyHbZAc3", "5Hddm3iBFD2GLT5ik7LZnT3XJUnRnN8PoeCFgGQgawUVKNm8", "5HEo565WAy4Dbq3Sv271SAi7syBSofyfhhwRNjFNSM2gP9M2", "5FcXnzNo3mrqReTEY4ftkg5iXRBi61iyvM4W1bywZLRqfxAY", "5HNQURvmjjYhTSksi8Wfsw676b4owGwfLR2BFAQzG7H3HhYf", "5FLKnbMjHY8LarHZvk2q2RY9drWFbpxjAcR5x8tjr3GqtU6F", "5Gpt8XWFTXmKrRF1qaxcBQLvnPLpKi6Pt2XC4vVQR7gqNKtU"])
         self.blockchain_sync_delta = self.get_config_value('blockchain_sync_delta', {'bitcoin': 100, 'doge': 100})
         self.grace_period = self.get_config_value('grace_period', False)
+        self.vali_penality_rate = self.get_config_value('vali_penalty_rate', 0.3)
```
Comment (aph5nt) [inline neurons/remote_config.py:151]: make default 0, so we can commit this now without affecting current status quo

**Review by aph5nt — CHANGES_REQUESTED** [review]

File: neurons/remote_config.py (hunk 4 above)
Comment (wendellswa06) [reply neurons/remote_config.py:151]: ok

File: neurons/remote_config.py (hunk 2 above)
Comment (wendellswa06) [reply neurons/remote_config.py:92]: we can use local subtensor

**Pull request merged**